lade/decoding.py

```diff
@@ -24,6 +25,30 @@ def greedy_search_proxy(self, *args, **kwargs):
     else:
         return FUNC_MAP["greedy_search"](self, *args, **kwargs)
 
+def sample_proxy(self, *args, **kwargs):
+    USE_LADE = int(os.environ.get("USE_LADE", 0))
+    CHAT = int(os.environ.get("CHAT", 0))
+    # There may be a more efficient way of extracting this
+    temperature = 0.0
+    for l in kwargs["logits_warper"]:
+        if isinstance(l, TemperatureLogitsWarper):
+            temperature = l.temperature
+            break
+    if CHAT and USE_LADE:
+        return jacobi_greedy_search_multilevel(self, chat=True, temperature=temperature, *args, **kwargs)
+    elif CHAT:
+        return greedy_search_chat(self, *args, **kwargs)
+    
+    if USE_LADE:
+        return jacobi_greedy_search_multilevel(self, chat=False, temperature=temperature, *args, **kwargs)
+    else:
+        return FUNC_MAP["sample"](self, *args, **kwargs)
+
+def gumbel(logits, tau):
+    gumbels = (
+        -torch.empty_like(logits, memory_format=torch.legacy_contiguous_format).exponential_().log()
+    )  # ~Gumbel(0,1)
+    return (logits + gumbels) / tau  # ~Gumbel(logits,tau)
```

Comment:
```suggestion
def gumbel(logits, tau):
    gumbels = (
        -torch.empty_like(logits, memory_format=torch.legacy_contiguous_format).exponential_().log_()
    )  # ~Gumbel(0,1)
    return gumbels.add_(logits).div_(tau)  # ~Gumbel(logits,tau)
```

doing it in place will be slightly more efficient (no need to create extra tensors)